### PR TITLE
Wrapper for workflow runs from files in workspace

### PR DESCRIPTION
**Pull request by kartikgupta-db**

### CODE REVIEW

File: packages/databricks-sdk-js/src/api-client.ts [hunk 1]
```diff
@@ -125,7 +125,10 @@ export class ApiClient {
 
         if (payload) {
             if (method === "POST") {
-                options.body = JSON.stringify(payload);
+                options.body =
```
Comment (fjakobs) [inline packages/databricks-sdk-js/src/api-client.ts:128]: why did you have to make this change?

File: packages/databricks-vscode/resources/python/file.workflow-wrapper.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        project_root = arg
+    else:
+        continue
+    databricks_arg_idx.extend([i, i-1])
+
+if python_file is None:
+    raise Exception("--databricks-source-file argument not specified")
+
+if project_root is None:
+    raise Exception("--databricks-project-root argument not specified")
```
Comment (fjakobs) [inline packages/databricks-vscode/resources/python/file.workflow-wrapper.py:23]: @pietern We need to remember to pass the arguments from DABs jobs.

File: packages/databricks-vscode/resources/python/notebook.workflow-wrapper.json [hunk 3]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (fjakobs) [inline packages/databricks-vscode/resources/python/notebook.workflow-wrapper.json:1]: Could we get rid of this file and create the JSON dynamically? the cell source should be identical to the workflow wrapper and it's easy to forget to keep them in sync.,

File: packages/databricks-vscode/resources/python/notebook.workflow-wrapper.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+#This file is autogenerated by the Databricks Extension for VS Code
```
Comment (fjakobs) [inline packages/databricks-vscode/resources/python/notebook.workflow-wrapper.py:1]: I know it's annoying but wouldn't we also have to do this for Scala and R?

File: packages/databricks-vscode/resources/python/notebook.workflow-wrapper.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+dbutils.widgets.text("DATABRICKS_SOURCE_FILE", "")
+dbutils.widgets.text("DATABRICKS_PROJECT_ROOT", "")
+
+if dbutils.widgets.get("DATABRICKS_SOURCE_FILE") != "":
+	import os
+	os.chdir(os.path.dirname(dbutils.widgets.get("DATABRICKS_SOURCE_FILE")))
+
+if dbutils.widgets.get("DATABRICKS_PROJECT_ROOT") != "":
+	import sys
+	sys.path.append(dbutils.widgets.get("DATABRICKS_PROJECT_ROOT"))
```
Comment (fjakobs) [inline packages/databricks-vscode/resources/python/notebook.workflow-wrapper.py:12]: in "file" mode you also append the dirname of the file.

File: packages/databricks-vscode/resources/python/file.workflow-wrapper.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+#remove databricks args from argv
+sys.argv = [value for i,value in enumerate(sys.argv) if i not in databricks_arg_idx]
+
+# change working directory
+os.chdir(os.path.dirname(python_file))
+
+# update python path
+sys.path.append(project_root)
+sys.path.append(os.path.dirname(python_file))
```
Comment (fjakobs) [inline packages/databricks-vscode/resources/python/file.workflow-wrapper.py:33]: This should already be set.

File: packages/databricks-vscode/src/run/WorkflowRunner.ts [hunk 7]
```diff
@@ -125,14 +126,38 @@ export class WorkflowRunner implements Disposable {
...
             }
         });
+
         try {
-            if (await isNotebook(program)) {
+            const notebookType = await isNotebook(program);
+            if (notebookType) {
+                let remoteFilePath: string =
+                    syncDestination.localToRemoteNotebook(program).path;
+                if (workspaceConfigs.enableFilesInWorkspace) {
```
Comment (fjakobs) [inline packages/databricks-vscode/src/run/WorkflowRunner.ts:135]: This branch should look at the sync destination path and not the setting. The user might still have a Repo path after changing the setting.

File: packages/databricks-vscode/src/run/WorkflowRunner.ts [hunk 8]
```diff
@@ -143,10 +168,21 @@ export class WorkflowRunner implements Disposable {
                     })
                 );
             } else {
+                const originalFile = syncDestination.localToRemote(program);
+                const wrappedFile = workspaceConfigs.enableFilesInWorkspace
```
Comment (fjakobs) [inline packages/databricks-vscode/src/run/WorkflowRunner.ts:172]: same as above

File: packages/databricks-vscode/src/workspace-fs/WorkspaceFsWorkflowWrapper.ts [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+        if (!WorkspaceFsUtils.isFile(wrappedFile)) {
+            throw new Error(
+                `Cannot create workflow wrapper for ${remoteFilePath.path}`
+            );
+        }
+        return wrappedFile;
+    }
+
+    @logging.withLogContext(Loggers.Extension)
+    private async createIpynbWrapper(
```
Comment (fjakobs) [inline packages/databricks-vscode/src/workspace-fs/WorkspaceFsWorkflowWrapper.ts:70]: this file needs unit tests

**Review by fjakobs — CHANGES_REQUESTED** [review]

File: packages/databricks-vscode/resources/python/notebook.workflow-wrapper.py (hunk 4 above)
Comment (kartikgupta-db) [reply packages/databricks-vscode/resources/python/notebook.workflow-wrapper.py:1]: This requires more investigation. I am not sure how module resolution works for scala and what the user expectations are. We can take this as a followup.

File: packages/databricks-vscode/resources/python/notebook.workflow-wrapper.py (hunk 5 above)
Comment (kartikgupta-db) [reply packages/databricks-vscode/resources/python/notebook.workflow-wrapper.py:12]: The dirname is already set currectly. It is redundant in the file mode. Removing from there.

File: packages/databricks-sdk-js/src/services/wsfs/WorkspaceFsDir.ts [hunk 10]
```diff
@@ -77,6 +79,8 @@ export class WorkspaceFsDir extends WorkspaceFsEntity {
         @context ctx?: Context
     ) {
         const validPath = this.getAbsoluteChildPath(path);
+        // eslint-disable-next-line no-console
+        console.log(validPath);
```
Comment (fjakobs) [inline packages/databricks-sdk-js/src/services/wsfs/WorkspaceFsDir.ts:83]: we should not use plain console.log from library code.

File: packages/databricks-sdk-js/src/services/wsfs/WorkspaceFsDir.ts [hunk 11]
```diff
@@ -47,11 +52,80 @@ export class WorkspaceFsDir extends WorkspaceFsEntity {
...
+
+    @withLogContext(ExposedLoggers.SDK)
+    async createFile(
+        path: string,
+        content: string,
+        overwrite = true,
+        @context ctx?: Context
+    ) {
+        const validPath = this.getAbsoluteChildPath(path);
+        // eslint-disable-next-line no-console
```
Comment (fjakobs) [inline packages/databricks-sdk-js/src/services/wsfs/WorkspaceFsDir.ts:82]:
```suggestion
```

File: packages/databricks-sdk-js/src/services/wsfs/WorkspaceFsDir.ts [hunk 12]
```diff
@@ -47,11 +52,80 @@ export class WorkspaceFsDir extends WorkspaceFsEntity {
...
+        console.log(validPath);
```
Comment (fjakobs) [inline packages/databricks-sdk-js/src/services/wsfs/WorkspaceFsDir.ts:83]:
```suggestion
```

File: packages/databricks-sdk-js/src/services/wsfs/WorkspaceFsDir.ts (hunk 10 above)
Comment (kartikgupta-db) [reply packages/databricks-sdk-js/src/services/wsfs/WorkspaceFsDir.ts]: removed

**Review by fjakobs — APPROVED** [review]

**Pull request merged**